src/services/formatting/rules.ts

```diff
@@ -300,8 +300,8 @@ namespace ts.formatting {
             rule("NoSpaceBeforeBinaryOperator", anyToken, binaryOperators, [isOptionDisabledOrUndefined("insertSpaceBeforeAndAfterBinaryOperators"), isNonJsxSameLineTokenContext, isBinaryOpContext], RuleAction.DeleteSpace),
             rule("NoSpaceAfterBinaryOperator", binaryOperators, anyToken, [isOptionDisabledOrUndefined("insertSpaceBeforeAndAfterBinaryOperators"), isNonJsxSameLineTokenContext, isBinaryOpContext], RuleAction.DeleteSpace),
 
-            rule("SpaceBeforeOpenParenInFuncDecl", anyToken, SyntaxKind.OpenParenToken, [isOptionEnabled("insertSpaceBeforeFunctionParenthesis"), isNonJsxSameLineTokenContext, isFunctionDeclContext], RuleAction.InsertSpace),
-            rule("NoSpaceBeforeOpenParenInFuncDecl", anyToken, SyntaxKind.OpenParenToken, [isOptionDisabledOrUndefined("insertSpaceBeforeFunctionParenthesis"), isNonJsxSameLineTokenContext, isFunctionDeclContext], RuleAction.DeleteSpace),
+            rule("SpaceBeforeOpenParenInFuncDecl", [SyntaxKind.Identifier, SyntaxKind.GreaterThanToken], SyntaxKind.OpenParenToken, [isOptionEnabled("insertSpaceBeforeFunctionParenthesis"), isNonJsxSameLineTokenContext, isFunctionDeclContext], RuleAction.InsertSpace),
+            rule("NoSpaceBeforeOpenParenInFuncDecl", [SyntaxKind.Identifier, SyntaxKind.GreaterThanToken], SyntaxKind.OpenParenToken, [isOptionDisabledOrUndefined("insertSpaceBeforeFunctionParenthesis"), isNonJsxSameLineTokenContext, isFunctionDeclContext], RuleAction.DeleteSpace),

```

Comment:
Previously, bugs in this line along with [the `SpaceAfterColon` above](https://github.com/microsoft/TypeScript/pull/41273/files#diff-7cb606e938a01498a467ff2d9e4715e413b9ed692137e0852998ef7357d9abaeR54) were coincidentally cancelling each other out, resulting in a forced "consistent" formatting.

This line (`NoSpaceBeforeOpenParenInFuncDecl`) used to remove spaces at:
```ts
function func(): (arg: any) => void { }
                ^here
```
since it's `left = anyToken` and `right = OpenParenToken`.
And `SpaceAfterColon` then overrides this with `left = ColonToken` and `right = anyToken` and inserts a space in between.
Since `isNotTypeAnnotationContext` is being added, the override is no longer applied here, and thus revealed the flaw at this line.

Now that `SpaceAfterColon` is no longer required as a high priority rule to override this, it may even be added as a configurable option along with `NotSpaceBeforeColon`, into something line `insertBefore/AfterColonInObjectLiteral`. But that's out of the topic of this PR.